### ISSUE TITLE
Cross-platform OpenCL misbehaviour

### ISSUE THREAD

**Issue by magick777** [genesismining/sgminer-gm#25]
**Edit: not obviously an issue with sgminer but with running it against amdgpu-pro 16.50**.

Opening this as a discrete issue, though it likely relates closely to some of the foregoing ones. Setting up a third rig, in Ubuntu initially and on 16.50, I get a flurry of HW errors and no useful work from a config that works on other machines. The problem is characterised by reporting ~2x the plausible hashrate for ethash but with errors; I have run into it on a couple of machines, but been more able to fix it on some than on others. The problem neither seems to lie squarely with the 16.50 driver nor clearly with 5.5.4, though it seems to connect with what has or has not been installed previously. That is to say:

- having had the same problem on another machine, it persisted with older versions of sgminer against the same driver, but in that case, rebuilding everything, removing old binaries and nuking anything sgminer-related in /usr/local/bin solved it, and on that machine I've 5.5.4 playing nicely with amdgpu-pro 16.50 under 16.04.1.

- however, I have as yet been incapable of fixing it on the new rig, the same in every respect as a rig that works except that it hasn't seen older versions of sgminer or amdgpu-pro in the same fashion. Packages currently installed are identical, same things tried, user baffled. Installing 5.4.0 didn't fix it, which pushes suspicion towards the drivers (I'll try downgrading). Machine has a working OpenCL setup and will run silentarmy for ZEC or genoil for ETH, pointing me back towards sgminer.

- now for the red herring. Same rig has a newly installed Windows 10, onto which I installed 16.12.2 drivers, AMDAPPSDK-3.0 and Windows binary of 5.5.4. Stock BIOS, a much simpler config, and the first thing that happens when I get sgminer running? The exact same behaviour from 5.5.4 on a different operating system and a different OpenCL platform.

Quite where that sites the problem, I don't know. In all cases I'm referring to ethash, my suspicion is that something is wrong in the new DAG handling code. Whether and how far this interacts with on-disk binaries, .cl sources and driver versions I'm not adequately certain; it interacts with *something* as I have 5.5.4 working against 16.50 on other rigs, including where it had previously done the same thing, but quite how I got it so I'm not able to reproduce.

Pursuant to issues reported by others, I'll try the 5.4.0 binary cleanly under Ubuntu before trying an older driver, and ditto under Windows. If there is a bug introduced in 5.5.4, that may help isolate it as such. So far, nothing is working too well for me, and a bitwise clone of another rig is starting to look like an attractive option...!

**Edit 1**: under Linux, the pre-compiled 5.4.0 binary showed the same behaviour, including when recompiling the binary having supplied the ethash.cl from that version. Doesn't obviously suggest that it relates to changes in 5.5.4. 

**Edit 2**: removing the 16.50 driver and installing 16.30 causes the same 5.4.0 binary to start working as expected. It begins to look like the 16.50 amdgpu-pro release is incomplete, in a way that shows up only on systems that haven't seen older versions.

- removing 16.30 and installing 16.40 did not cause anything to break, for the same 5.4.0 binary
- removing 16.40 and installing 16.50 **caused the same issue** to recur for the same 5.4.0 binary
     - and likewise for 5.5.4 compiled from a clean git checkout
    - despite that I have it working with the same driver on an identical system
        - copying the working sgminer executable from that system, the very same issue
        - that system has an ethashEllesmeregw256l8.bin dating from Nov 26th, size 253264 bytes
            - copying the older .bin file to the offending system is what causes it to "work" with 16.50

Now we're getting somewhere; **the problem seems to be with .bin files compiled by the 16.50 driver** (which brings a dependency on libllvm-3.8, compiler version bump on AMD's part, and would apply to Linux and Windows releases). The notion of "it works with 16.50" may be "until you compile something".

**Edit 3:** to complete the picture, I reverted to the 16.40 driver and allowed 5.5.4 to compile a .bin from its own sources. This worked fine, producing a .bin 32 bytes larger than the one from 5.4.0 and 38kB smaller than the .bin that's compiled by the 16.50 driver from the same code. Pretty clear now where the issue lies.

**Comment by tpruvot** [comment]
its a driver issue, they migrated the opencl compiler lib to a new langage level, with a new version of CLANG, if you just replace the amdocl64 lib with the previous one, it should work... same on both win and linux. But, seems to only affect ethash here. I didnt dig a lot, just have a RX480 to follow evolutions

**Comment by magick777** [comment]
Yes, that's what I figured - eventually! The default of keeping the .bin files can mask where a problem creeps in - I wonder if the filenames should also contain the version of the driver for which they were compiled.

Replacing the ICD library is a neat workaround if that's the only issue. However, would a change to that library account for this difference in the size of the compiled binaries?

```
-rw-r--r-- 1 root root 253296 Jan  8 19:44 ethashEllesmeregw256l8.bin
-rw-r--r-- 1 root root 292176 Jan  8 18:45 ethashEllesmeregw256l8.bin.old
```

In the above, the first is compiled by the 16.40 driver from 5.5.4 sources, while .old is the same thing compiled by the 16.50 driver. I've no idea what the extra 38K on the .bin is about. 

Am I the only one thinking it would make sense to implement Vulkan and a SPIR-V toolchain as compared with enduring AMD's OpenCL development?

**Comment by joker5bb** [comment]
How can I use amdgpu kernel drivers?

**Comment by jramos** [comment]
This just bit me, too. I was inexplicably getting:

```
traps: cg@2/Miner[1348] general protection ip:7ff792cabe79 sp:7ff7577fdba0 error:0 in libamdocl64.so[7ff7925f5000+3db6000]
```

Removing the old .bin files did the trick.

**Comment by OhGodAGirl** [comment]
This bug is fixed by using ethash-new; 16.40 and up! This will only work for worksizes less than or equal to 64.
